src/ser.rs

```diff
@@ -24,17 +61,89 @@ pub trait PythonizeListType: Sized {
         U: ExactSizeIterator<Item = T>;
 }
 
+// TODO: remove 'py lifetime once GATs are available in MSRV
 /// Custom types for serialization
-pub trait PythonizeTypes {
+pub trait PythonizeTypes<'py> {
     /// Python map type (should be representable as python mapping)
-    type Map: PythonizeDictType;
+    type Map: PythonizeMappingType<'py>;
+    /// Python (struct-like) named map type (should be representable as python mapping)
+    type NamedMap: PythonizeNamedMappingType<'py>;
     /// Python sequence type (should be representable as python sequence)
     type List: PythonizeListType;
 }
 
-impl PythonizeDictType for PyDict {
-    fn create_mapping(py: Python) -> PyResult<Bound<PyMapping>> {
-        Ok(PyDict::new_bound(py).into_any().downcast_into().unwrap())
+impl<'py> PythonizeMappingType<'py> for PyDict {
+    type Builder = Bound<'py, Self>;
+
+    fn builder(py: Python<'py>, _len: Option<usize>) -> PyResult<Self::Builder> {
+        Ok(Self::new_bound(py))
+    }
+
+    fn push_item(
+        builder: &mut Self::Builder,
+        key: Bound<'py, PyAny>,
+        value: Bound<'py, PyAny>,
+    ) -> PyResult<()> {
+        builder.set_item(key, value)
+    }
+
+    fn finish(builder: Self::Builder) -> PyResult<Bound<'py, PyMapping>> {
+        Ok(builder.into_mapping())
+    }
+}
+
+/// Adapter type to use an unnamed mapping type, i.e. one that implements
+/// [`PythonizeMappingType`], as a named mapping type, i.e. one that implements
+/// [`PythonizeNamedMappingType`]. The adapter simply drops the provided name.
+///
+/// This adapter is commonly applied to use the same unnamed mapping type for
+/// both [`PythonizeTypes::Map`] and [`PythonizeTypes::NamedMap`] while only
+/// implementing [`PythonizeMappingType`].
+pub struct PythonizeUnnamedMappingAdapter<'py, T: PythonizeMappingType<'py>> {
+    unnamed: T,
+    _marker: PhantomData<&'py ()>,
+}
+
+impl<'py, T: PythonizeMappingType<'py>> PythonizeUnnamedMappingAdapter<'py, T> {
+    #[must_use]
+    pub fn new(unnamed: T) -> Self {
+        Self {
+            unnamed,
+            _marker: PhantomData::<&'py ()>,
+        }
+    }
+
+    #[must_use]
+    pub fn into_inner(self) -> T {
+        self.unnamed
+    }
+}
+
+impl<'py, T: PythonizeMappingType<'py>> From<T> for PythonizeUnnamedMappingAdapter<'py, T> {
+    fn from(value: T) -> Self {
+        Self::new(value)
+    }
+}
```

Comment:
Coverage is suggesting that these lines are unused; as I think this struct is a type-level only feature it probably is correct that these can be removed?

```suggestion
```